feat(outcome): Emit outcomes for rejected events

RaduW opened the pull request:
This is work in progress, let's have a look at it to check that I'm on the right track.

untitaker committed:
release: 0.4.44

RaduW committed:
feat(outcome): WIP added outcome actor and started to send outcomes

RaduW committed:
feat(outcome): WIP cleanup OutcomeReason

RaduW committed:
feat(outcome): WIP use clone! macro instead of manually cloning and moving

RaduW committed:
feat(outcome): WIP finished handling outcomes in store endpoint

untitaker committed:
Merge release/0.4.44 into master

untitaker committed:
feat: Add schema for quota config (#235)

* feat: Add schema for quota config

* fix: Fix tests

* fix: Set correct kind of key id

* fix: Fix tests

* fix: Address review comments

* ref: Log error for unreachable state

* ref: Remove unneeded return statement

RaduW committed:
feat(outcome): Added outcome handling for event processing

jan-auer commented on server/src/utils/date_time.rs:
```diff
@@ -0,0 +1,15 @@
+use std::time::{Instant, SystemTime};
+
+/// Converts an Instant into a SystemTime.
+pub fn instant_to_system_time(instant: &Instant) -> SystemTime {
```
`Instant` implements `Copy`, so you can pass by value here. Clippy will most likely complain about this.

RaduW replied:
ok, done

jan-auer commented on general/src/filter/common.rs:
```diff
@@ -98,7 +98,8 @@ impl<'de> Deserialize<'de> for GlobPatterns {
 /// (e.g.  "browser-extensions")
-#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Serialize, Hash)]
+#[serde(rename_all = "kebab-case")]
```
With the discussed changes, we can likely revert all changes in this file. It's no longer required to serialize this directly.

If serialization is still required in the end, please implement it using `.as_str()` rather than deriving it a second time.

RaduW replied:
ok, done

jan-auer commented on server/src/actors/outcome.rs:
```diff
@@ -0,0 +1,312 @@
+    pub event_id: Option<EventId>,
+}
+
+impl Message for KafkaOutcomeMessage {
```
Can we decouple the interface of the outcomes actor from the internal serialization to JSON? That would allow us to create a cleaner interface for the outcomes enum, by merging the outcome and the reason into one enum.

RaduW replied:
ok, done

jan-auer commented on server/src/actors/project.rs:
```diff
@@ -429,7 +434,8 @@ impl Handler<GetEventAction> for Project {
             // Compensate for the missing subsec part by adding 1s
             let secs = (self.retry_after - now).as_secs() + 1;
-            return Response::ok(EventAction::RetryAfter(secs));
+            let reason = "TODO_ADD_REASON".to_string(); // TODO we need to also keep the reason
```
Please keep the initial reason why this project was rate limited. Then, attach the reason instead of creating a new one.

For follow-up: We need to remember rate limits per project key separately. Please keep a map from public key to `(Instant, String)` for this.

jan-auer commented on server/src/actors/project.rs:
```diff
@@ -405,12 +410,12 @@ impl GetEventAction {
 pub enum EventAction {
     /// The event should be discarded.
-    Discard,
+    Discard(OutcomeReason),
```
Please check if this can have just an `OutcomeInvalidReason`. We can consider to rename it to something like `DiscardReason`. This would lead to a variant `Outcome::Invalid(DiscardReason)`, which might be acceptable.

RaduW replied:
Ok, done

jan-auer commented on server/src/actors/events.rs:
```diff
@@ -1,4 +1,4 @@
-use std::sync::Arc;
+use std::sync::{Arc, Mutex};
```
Please use [`parking_lot::Mutex`](https://docs.rs/parking_lot/0.9.0/parking_lot/type.Mutex.html) instead.

RaduW replied:
ok, done

jan-auer commented on server/src/actors/events.rs:
```diff
@@ -94,13 +79,16 @@ enum ProcessingError {
     Shutdown,
+
+    #[fail(display = "upstream rate limited")]
+    UpstreamRateLimited,
```
Please keep this merged with `RateLimited(u64, String)`.

RaduW replied:
ok, done

jan-auer commented on server/src/actors/events.rs:
```diff
@@ -550,6 +559,48 @@ impl Handler<HandleEvent> for EventManager {
                 log::warn!("error processing event {}: {}", event_id, LogError(&error));
                 metric!(counter("event.rejected") += 1);
+                let outcome_params: Option<(Outcome, OutcomeReason)> = match &error {
+                    ProcessingError::InvalidJson(_)
```
This is kind-of special. It's not an internal failure if the JSON payload from a client is invalid. This should have an `InvalidJson` reason, similar to when the payload is too large.

RaduW replied:
Ok, done

jan-auer commented on general/src/reason.rs:
```diff
@@ -0,0 +1,96 @@
+    Cors,
+
+    //new errors emitted by Semaphore (and not found in Sentry)
+    StorePayloadError,
```
Can we split this into more concrete types:

```
InvalidJson,
UnknownLength,
InvalidCompression,
```

RaduW replied:
ok, done

tonyo committed:
fix: Remove Freight k8s config

untitaker committed:
fix: Actually restore datascrubbing defaults (#237)

* fix: Actually restore datascrubbing defaults

https://github.com/getsentry/semaphore/pull/227/commits/71531c006841d8218c6fda406cfddd66c8aa9263 was incorrectly committed -- it intended to disable PII stripping by default, but actually didn't.

* test: Add test

RaduW committed:
feat(outcome): separated actor outcome message from kafka outcome message

untitaker committed:
fix: Fix entirely broken geoip lookup (#240)

mitsuhiko committed:
meta: changelog for 0.4.45

mitsuhiko committed:
release: 0.4.45

mitsuhiko committed:
Merge release/0.4.45 into master

untitaker commented on common/src/macros.rs:
```diff
@@ -61,3 +61,40 @@ macro_rules! tryf {
+
+/// An alternative to a `move` closure.
+///
+/// When one needs to use a clojure with move semantics one often needs to clone and
```
```suggestion
/// When one needs to use a closure with move semantics one often needs to clone and
```

untitaker commented on server/src/service.rs:
```diff
@@ -102,24 +103,31 @@ pub struct ServiceState {
-        let event_manager = EventManager::create(config.clone(), upstream_relay.clone())
-            .context(ServerErrorKind::ConfigError)?
-            .start();
+        let outcome_producer = OutcomeProducer::create(config.clone())?.start();
```
This is inconsistent with how StoreForwarder works, because `OutcomeProducer` resolves to a noop internally while StoreForwarder is stored as `Option<>` and only instantiated if processing is enabled. I would like to do the same for OutcomeProducer because I hope to have a single place in `EventManager::create` that does the check for `Is_processing_enabled` only once. The other reason is that I already applied StoreForwarder's idiom to quotas as well.

Having the callsite of `track_outcome` check for the `None`-ness of your `Option<Addr<OutcomeProducer>>` will also help avoid creating unnecessary structs and strings when processing is disabled. I'd also argue that `if let Some(producer) = outcome_producer { }` everywhere is not as much code as having the noop implementation

RaduW replied:
Yes you are right, it is inconsistent with the way StoreForwarder was implemented. We need to decide if it is worthwhile changing the implementation of OutcomeProducer for this.

jan-auer replied:
Discussed offline. In this case we agree that we use a different pattern. In fact, `OutcomeProducer` will start to emit a debug log for every outcome now, and may potentially emit metrics in the future. 

Thanks for raising this @untitaker!

untitaker commented on server/src/actors/outcome.rs:
```diff
@@ -0,0 +1,312 @@
+
+    #[test]
+    fn outcome_serialization_test() {
+        let val = serde_json::to_string(&Outcome::Accepted).unwrap_or("error".to_string());
```
What does this test do?

RaduW replied:
Nothing, used to see how a serialised message looks, left there by mistake. Thanks for noticing it.

untitaker commented on server/src/actors/outcome.rs:
```diff
@@ -0,0 +1,312 @@
+    }
+}
+
+//impl Serialize for Outcome {
```
Dead code

untitaker reviewed (approved):
I have no strong opinions on what I wrote and believe this is ready for deployment anyway.

RaduW committed:
feat(outcome): fixes from code review

untitaker committed:
fix: Clean up ip address normalization again (#242)

* fix: Clean up ip address normalization again

* fix: Fix tests

* fix: Actual line-by-line port of python

untitaker committed:
doc: Changelog for 0.4.46

untitaker committed:
release: 0.4.46

jan-auer commented on server/src/actors/outcome.rs:
```diff
@@ -0,0 +1,416 @@
+
+/// Message
+#[derive(Debug)]
+pub struct OutcomeMessage {
```
_nit_: Can we call this `TrackOutcome`, so that it is named similar to Sentry's `track_outcome` function?

RaduW replied:
ok, done

jan-auer commented on server/src/actors/outcome.rs:
```diff
@@ -0,0 +1,416 @@
+pub use self::no_op_implementation::*;
+use crate::body::StorePayloadError;
+use crate::endpoints::store::BadStoreRequest;
+use semaphore_general::filter::FilterStatKey;
```
_nit_: Please move imports to the top.

RaduW replied:
ok, done

jan-auer commented on server/src/actors/outcome.rs:
```diff
@@ -0,0 +1,416 @@
+
+impl Outcome {
+    /// Returns the name of the outcome as recognized by Sentry.
+    pub fn name(&self) -> &'static str {
```
This methods reports as unused without the `processing` feature. Please add `#[cfg(feature = "processing")]`.

RaduW replied:
ok, done

jan-auer commented on server/src/utils/date_time.rs:
```diff
@@ -0,0 +1,15 @@
+    SystemTime::now() - instant.elapsed()
+}
+
+pub fn instant_to_unix_timestamp(instant: Instant) -> u64 {
```
This method is only used with the processing feature. Please either add `#[cfg(feature = "processing")]` (preferred) or `#[allow(unused)]`. 

Also, can we please add a doc comment?

RaduW replied:
ok, done

jan-auer commented on server/src/actors/outcome.rs:
```diff
@@ -0,0 +1,416 @@
+                // when processing is disabled we don't send outcomes
+                return Box::new(futures::future::ok(()));
+            }
+            if let Some(ref producer) = self.producer {
```
_nit_: The flag `self.is_enabled` and the option `self.producer` carry the same information. You could remove the flag and just unwrap the producer like this:

```rust
let producer = match self.producer {
    Some(ref producer) => producer,
    None => return Box::new(futures::future::ok(())),
};

// ...
```

RaduW replied:
ok, done

jan-auer commented on server/src/actors/outcome.rs:
```diff
@@ -0,0 +1,416 @@
+                let payload = tryf![serde_json::to_string(&OutcomePayload::from(&message))
+                    .map_err(OutcomeError::SerializationError)];
+
+                metric!(counter("events.outcomes") +=1 , "reason"=>message.outcome.to_reason().unwrap_or(""), 
```
_nit_: This line exceeds maximum line length. Unfortunately, `rustfmt` doesn't format macros. Please break line before `"reason"`.

RaduW replied:
ok, done

jan-auer commented on server/src/actors/outcome.rs:
```diff
@@ -0,0 +1,416 @@
+                metric!(counter("events.outcomes") +=1 , "reason"=>message.outcome.to_reason().unwrap_or(""), 
+                    "outcome"=>message.outcome.name() );
+
+                // Since eventId is optional and we wa
```
_nit_: Incomplete comment?

RaduW replied:
yes

jan-auer commented on server/src/actors/outcome.rs:
```diff
@@ -0,0 +1,416 @@
+    impl Handler<OutcomeMessage> for OutcomeProducer {
+        type Result = ResponseFuture<(), OutcomeError>;
+
+        fn handle(&mut self, message: OutcomeMessage, _ctx: &mut Self::Context) -> Self::Result {
```
_nit_: What if we emit a debug log here? This could be nice to debug Relay - even if processing is disabled - to see where events go. We can even do this in the noop implementation, because it might be valuable information.

RaduW replied:
ok, done

jan-auer commented on server/src/actors/outcome.rs:
```diff
@@ -0,0 +1,416 @@
+    use crate::utils::{SyncFuture, SyncHandle};
+    use failure::{Fail, ResultExt};
+    use semaphore_common::tryf;
+    use semaphore_common::{metric, KafkaTopic, Uuid};
```
_nit_: Please sort imports: `super` -> _external crates_ -> `semaphore_*` -> `crate::*`

RaduW replied:
ok, done

jan-auer commented on server/src/actors/project.rs:
```diff
@@ -337,7 +338,7 @@ impl ProjectState {
         // Try to verify the request origin with the project config.
         if !self.is_valid_origin(meta.origin()) {
-            return EventAction::Discard;
+            return EventAction::Discard(DiscardReason::ProjectId);
```
👏

jan-auer commented on server/src/endpoints/store.rs:
```diff
@@ -149,8 +181,19 @@ fn store_event(
                         .map_err(BadStoreRequest::ScheduleFailed)
                         .and_then(|result| result.map_err(BadStoreRequest::ProcessingFailed))
+                        .map_err(move |e| {
+                            outcome_producer.do_send(OutcomeMessage {
```
_nit_: We have 4 calls to outcome producer in this chain of futures. Each of those calls happens before emitting an error. Would it be possible to move this all the way to the bottom so that there is only a single call?

jan-auer commented on server/src/actors/events.rs:
```diff
@@ -78,9 +63,9 @@ enum ProcessingError {
 
     #[fail(display = "event submission rejected")]
-    EventRejected,
+    EventRejected(DiscardReason),
```
_nit_: Add the reason to the message?

RaduW replied:
ok, done

jan-auer commented on server/src/actors/events.rs:
```diff
@@ -456,24 +453,32 @@ impl Handler<HandleEvent> for EventManager {
                         .map_err(ProcessingError::ScheduleFailed)
                         .and_then(|result| result.map_err(ProcessingError::PiiFailed))))
-                    .and_then(clone!(meta, |project_state| processor
+                    .and_then(clone! (meta,  |project_state| {
```
_nit_: Seems like there are too many spaces here that are not formatted by `rustfmt` due to the macro.

RaduW replied:
ok, done

jan-auer commented on server/src/actors/events.rs:
```diff
@@ -550,6 +557,47 @@ impl Handler<HandleEvent> for EventManager {
+                        Some(Outcome::Invalid(DiscardReason::Internal))
+                    }
+                    ProcessingError::EventRejected(outcome_reason) => {
+                        Some(Outcome::Invalid(*outcome_reason))
```
_nit_: Match directly on `error`. `OutcomeReason` implements `Copy`, so you do not need the implicit ref here. Same applies to `FilterStatKey` below. You will need to match `RateLimited(_timeout, ref reason)` below, though.

RaduW replied:
ok, done

jan-auer commented on server/src/actors/events.rs:
```diff
@@ -550,6 +557,47 @@ impl Handler<HandleEvent> for EventManager {
+                        Some(Outcome::Filtered(*filter_stat_key))
+                    }
+
+                    ProcessingError::NoAction(_)
```
Please map `NoAction` to `Outcome::Internal`. It indicates a failure to determine which action to apply to an event.

RaduW replied:
ok, done

untitaker committed:
Merge release/0.4.46 into master

RaduW committed:
Update common/src/macros.rs

Co-Authored-By: Markus Unterwaditzer <markus@unterwaditzer.net>

RaduW committed:
feat(outcome): Renamed OutcomeMessage to TrackOutcome

mitsuhiko committed:
feat: Added event capture mode for unittesting (#238)

RaduW committed:
feat(outcome): Code review fixes

jan-auer commented on server/src/endpoints/store.rs:
```diff
@@ -44,19 +44,22 @@ pub enum BadStoreRequest {
 
         match self {
-            BadStoreRequest::RateLimited(secs) => {
+            BadStoreRequest::RateLimited(secs, _reason) => {
```
Can we add this reason to a response header, like `X-Sentry-Error` or `Warning`?

jan-auer reviewed (approved):
G2G after resolving conflicts with `master`. One minor comment attached.

RaduW committed:
Merge branch 'feat(kafka-outcome)'

# Conflicts:
#	server/src/actors/events.rs

RaduW committed:
fix clippy warnings
